virttest/qemu_devices/qcontainer.py

```diff
@@ -2853,3 +2856,93 @@ def fs_define_by_params(self, name, params, bus=None):
         else:
             raise ValueError("unsupported filesystem driver type")
         return devices
+
+    def tpm_define_by_params(self, name, params):
+        """
+        Create TPM devices by params.
+
+        :param name: Object name of TPM device.
+        :type name: str
+        :param params: TPM device params.
+        :type params: dict
+        :return: List of corresponding TPM devices.
+        :rtype: list
+        """
+        model = params.get("tpm_model")
+        backend = params.get("tpm_backend")
+        version = params.get("tpm_version")
+
+        tmp_dir = data_dir.get_tmp_dir()
+
+        def _handle_log(line):
+            try:
+                utils_misc.log_line('%s_%s_swtpm_setup.log' % (self.vmname, name), line)
+            except Exception as e:
+                logging.warn("Can't log %s_%s_swtpm_setup output: %s.", self.vmname, name, e)
+
+        def _emulator_setup(binary, extra_options=None):
+            setup_cmd = binary
+            if version in ('2.0',):
+                setup_cmd += ' --tpm2'
+
+            tpm_path = os.path.join(tmp_dir, '%s_%s_tpm_state' % (self.vmname, name))
+            if os.path.exists(tpm_path):
+                shutil.rmtree(tpm_path, ignore_errors=True)
+            os.makedirs(tpm_path)
+            setup_cmd += " --tpm-state %s" % tpm_path
+
+            setup_cmd += (" --createek --create-ek-cert"
+                          " --create-platform-cert"
+                          " --lock-nvram --not-overwrite")
+
+            if extra_options:
+                setup_cmd += extra_options
+
+            logging.info('Running TPM emulator setup command: %s.', setup_cmd)
+            _process = aexpect.run_bg(setup_cmd, None, _handle_log, auto_close=False)
+            status_ending = 'Ending vTPM manufacturing'
+            _process.read_until_any_line_matches(status_ending, timeout=5)
+            return tpm_path
+
+        devices = []
+        if backend == 'emulator':
+            tpm_bin = params.get("tpm_bin", "/usr/bin/swtpm")
+            tpm_setup_bin = params.get("tpm_setup_bin", "/usr/bin/swtpm_setup")
```

Comment:
Also here if `/usr/bin/swtpm_setup` exists should be checked.

Reply:
Please refer to https://github.com/avocado-framework/avocado-vt/pull/2673#discussion_r468432673
Thanks.

Reply:
Hi, @vivianQizhu 
Have updated, please help me to review.
Thanks.